source/extensions/common/wasm/wasm.h

```diff
@@ -183,6 +182,34 @@ void clearCodeCacheForTesting();
 void setTimeOffsetForCodeCacheForTesting(MonotonicTime::duration d);
 WasmEvent toWasmEvent(const std::shared_ptr<WasmHandleBase>& wasm);
 
+class PluginConfig : Logger::Loggable<Logger::Id::wasm> {
+public:
+  PluginConfig(const envoy::extensions::wasm::v3::PluginConfig& config,
+               Server::Configuration::ServerFactoryContext& context, Stats::Scope& scope,
+               Init::Manager& init_manager, envoy::config::core::v3::TrafficDirection direction,
+               const envoy::config::core::v3::Metadata* metadata, bool singleton);
+
+  std::shared_ptr<Context> createContext();
+  Wasm* wasmOfHandle();
+  const PluginSharedPtr& plugin() { return plugin_; }
+
+private:
+  PluginSharedPtr plugin_;
+  RemoteAsyncDataProviderPtr remote_data_provider_;
+  const bool is_singleton_handle_{};
+
+  bool plugin_handle_initialized_{};
```

Comment:
WDYT about a variant with mono state, singleton handle, and thread local handle? Might be easier to understand the three states.

Reply:
SGTM.

Reply:
+1